src/Step/Deploy/InstallUpdate/Update/Setup.php

```diff
@@ -92,22 +94,41 @@ public function __construct(
     public function execute()
     {
         $this->flagManager->delete(FlagManager::FLAG_REGENERATE);
+        $verbosityLevel = $this->stageConfig->get(DeployInterface::VAR_VERBOSE_COMMANDS);
+        $installUpgradeLog = $this->fileList->getInstallUpgradeLog();
 
+        $outdated = false;
         try {
-            $verbosityLevel = $this->stageConfig->get(DeployInterface::VAR_VERBOSE_COMMANDS);
-            $installUpgradeLog = $this->fileList->getInstallUpgradeLog();
+            $this->logger->info('Checking if setup:upgrade is required');
 
-            $this->logger->info('Running setup upgrade.');
+            $this->shell->execute('php ./bin/magento setup:db:status --ansi --no-interaction ');
 
-            $this->shell->execute('echo \'Updating time: \'$(date) | tee -a ' . $installUpgradeLog);
-            $this->shell->execute(sprintf(
-                '/bin/bash -c "set -o pipefail; %s | tee -a %s"',
-                'php ./bin/magento setup:upgrade --keep-generated --ansi --no-interaction ' . $verbosityLevel,
-                $installUpgradeLog
-            ));
+            $this->logger->info('Upgrade check complete');
         } catch (\RuntimeException $exception) {
-            //Rollback required by database
-            throw new StepException($exception->getMessage(), 6, $exception);
+            if ($exception->getCode() == DbStatusCommand::EXIT_CODE_UPGRADE_REQUIRED) {
+                $outdated = true;
+            } else {
+                throw new ProcessException($exception->getMessage(), 6, $exception);
+            }
+        }
+
+        if ($outdated) {
```

Comment:
We change default behavior of upgrade that can cause issues.I suggest making this functionality configurable and disabled by default

Reply:
So perhaps adding something like :
```
# SKIP_SETUP_UPGRADE - skips setup:upgrade during the deploy phase if not required.                                           #
# Magento Version: 2.3.3 and later                                                                                    #
# Default value: false                                                                                              #
# Possible values: true or false                                                                                      #
# Stages: deploy                                                                                    #
# Example:                                                                                                            #
#          stage:                                                                                                     #
#            deploy:                                                                                                   #
#              SKIP_SETUP_UPGRADE: true       
```
In .magento.env.yaml ?

Reply:
Yep
Also I propose to name it more understandable, like
skip_upgrade_unless_required

Reply:
Sure!
I'm sure I can work out how to do this. I guess because of this I will need to add at least one Unit Test to Magento\MagentoCloud\Test\Unit\Step\Deploy\InstallUpdate\Update\SetupTest .

I have a question - as the code structure of ECE-TOOLS is different in the repo - are there some guidelines somewhere how run locally to test? Both the actual code and the UTs?

Reply:
@BarnyShergold you can run all functional test using Cloud Docker: https://devdocs.magento.com/guides/v2.3/cloud/docker/docker-development-testing.html

Unit tests can be run directly I think.

Reply:
Hi - Okay that might be fun then - I don't use Docker - Vagrant instead.

Reply:
@BarnyShergold , this actually just requires Docker itself to be installed - then following documentation everything will be deployed for you automatically.

Reply:
I've tried that but getting errors all over the place. What's the best place to go to discuss this and get some help?

Reply:
@BarnyShergold , please join https://magentocommeng.slack.com/ channel - #cloud. We can discuss it over Slack there.

Thanks!

Reply:
@BarnyShergold 
Thanks so much for your pull request.
Are you working on this PR?
If you want, we can finish this PR on our side.

Reply:
Hi Had to put to one side for now but was going to follow Kandy's suggestion and make it configurable - will try to pick up in next few days

Reply:
@arhiopterecs - All code changes made & commit with extensive Unit Tests added. Not sure why Travis is failing but it looks external to this module